src/compiler/checker/lines_12000_16000.rs

```diff
@@ -390,7 +400,21 @@ impl TypeChecker {
             if false {
                 unimplemented!()
             } else {
-                type_set.push(type_.type_wrapper());
+                let len = type_set.len();
+                let index: isize = if len > 0 && type_.id() > type_set[len - 1].id() {
+                    !isize::try_from(len).unwrap()
+                } else {
+                    binary_search_copy_key(
```

Comment:
The `binary_search()`/`insert_sorted()`/`binary_search_key()` functions were where I was running into lifetime issues - I didn't really end up using the `for<'a> ...` (higher-kinded lifetimes?) technique from [this answer](https://stackoverflow.com/a/47126516/732366) but did sort of do what it's saying as far as "hardcoding" the fact that the "key selector" closure returns a reference in `binary_search()`/`binary_search_key()`, so exposed parallel helpers `binary_search_copy_key()`/`binary_search_key_copy_key()` for when the key selector returns a `Copy`-able value rather than a reference. I could maybe revisit this since I think it's possible I don't even need to hardcode the reference-returning-ness in `binary_search()`/`binary_search_key()`

So here since the key for binary search is a `TypeId` (which is `Copy`), I'm using `binary_search_copy_key()`